bugfix: don't fetch package metadata for unknown concrete specs

tgamblin opened the pull request:
Fixes #36847.
Fixes #34108.

This fixes two issues:
1. The concretizer can fail with `reuse:true` if a buildcache or installation contains a package with a dependency that has been renamed or deleted in the main repo (e.g., `netcdf` was refactored to `netcdf-c`, `netcdf-fortran`, etc., but there are still binary packages with dependencies called `netcdf`).

    We should still be able to install things for which we are missing `package.py` files.

    `Spec.inject_patches_variant()` was failing this requirement by attempting to look up the package class for concrete specs.  This isn't needed -- we can skip it.

2. `spec_clauses()` attempts to look up package information for concrete specs in order to determine which virtuals they may provide. This fails for renamed/deleted dependencies of buildcaches and installed packages.

    This will eventually be fixed by https://github.com/spack/spack/pull/35258, but we need a workaround to make older buildcaches usable.

- [x] swap two conditions in `Spec.inject_patches_variant()`
- [x] make an exception for renamed packages and omit their virtual constraints
- [x] add a note that this will be solved by adding virtuals to edges

tgamblin committed:
bugfix: don't look up patches from packages for concrete specs

The concretizer can fail with `reuse:true` if a buildcache or installation contains a
package with a dependency that has been renamed or deleted in the main repo (e.g.,
`netcdf` was refactored to `netcdf-c`, `netcdf-fortran`, etc., but there are still
binary packages with dependencies called `netcdf`).

We should still be able to install things for which we are missing `package.py` files.

`Spec.inject_patches_variant()` was failing this requirement by attempting to look up
the package class for concrete specs.  This isn't needed -- we can skip it.

- [x] swap two conditions in `Spec.inject_patches_variant()`

tgamblin commented:
@eugeneswalker: does this fix your issue?

eugeneswalker commented:
This PR does indeed resolve https://github.com/spack/spack/issues/36847. Thank you!

tgamblin committed:
bugfix: don't look up virtual information for unknown packages

`spec_clauses()` attempts to look up package information for concrete specs in order to
determine which virtuals they may provide. This fails for renamed/deleted dependencies
of buildcaches and installed packages.

This will eventually be fixed by #35258, which adds virtual information on edges, but we
need a workaround to make older buildcaches usable.

- [x] make an exception for renamed packages and omit their virtual constraints
- [x] add a note that this will be solved by adding virtuals to edges

tgamblin commented:
@eugeneswalker see if the second commit fixes https://github.com/spack/spack/issues/34108.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1566,9 +1566,20 @@ class Body(object):
-                            # Ensure Spack will not coconcretize this with another provider
-                            # for the same virtual
-                            for virtual in dep.package.virtuals_provided:
+                            # Skip virtual node constriants for renamed/deleted packages,
```
```suggestion
                            # Skip virtual node constraints for renamed/deleted packages,
```

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1566,9 +1566,20 @@ class Body(object):
-                            for virtual in dep.package.virtuals_provided:
+                            # Skip virtual node constriants for renamed/deleted packages,
+                            # so their binaries can still be installed.
+                            # NOTE: with current specs (which lack edge attributes) this
+                            # can allow concretizations with two providers, but it's unlikely.
```
This is a bit scary :worried: Not sure what we could do more than this though. I agree it's unlikely, since one of the provider must be in a buildcache and not in the current repository.

tgamblin replied:
I think we don't rename things (especially MPI and other virtuals) frequently enough for this to matter.  Though @eugeneswalker and the E4S team are trying hard to make it matter with their 100,000 package buildcache.

Goal should be to get virtual info on edges before it matters 😅

tgamblin replied:
I think the value of reuse here outweighs the corner case.

tgamblin committed:
bugfix: allow reuse of packages from foreign namespaces

We currently throw a nasty error if you try to reuse packages from some other namespace
(e.g., OLCF), but we should be able to reuse patched local versions of builtin packages.

Right now the only obstacle to that is that we try to look up virtual info for unknown
namespaces, and we can't get the package from the repo to do that. We *can* assume that
a package with a known namespace is similar, and that its virtual provider information
is reasonably accurate, so we now do that. This isn't 100% accurate, but neither is
relying on the package itself, as it may have gone out of date.

The real solution here is virtual edge information, but this is a stopgap until we have
that.
